src/Plugin/Commerce/PaymentGateway/OnsiteBase.php

```diff
@@ -643,4 +665,289 @@ protected function remoteCustomerIdProviderKey() {
+   */
+  protected function buildAuthenticationFormElement() {
+    $form = [];
+
+    $form['authentication'] = [
+      '#type' => 'details',
+      '#title' => $this->t('Authentication'),
+      '#open' => TRUE,
+    ];
+    $form['authentication']['username'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Username'),
+      '#default_value' => $this->configuration['username'],
+      '#required' => TRUE,
+    ];
+    $form['authentication']['password'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Password'),
+      '#default_value' => $this->configuration['password'],
+      '#required' => TRUE,
+    ];
+
+    return $form;
+  }
+
+  /**
+   * Returns the form element that contains the statement descriptors fields.
+   *
+   * @return array
+   *   The statement descriptors form element render array.
+   */
+  protected function buildDescriptorFormElement() {
+    $form = [];
+    $config = &$this->configuration['statement_descriptors'];
+    $token_types = ['commerce_order', 'commerce_payment', 'commerce_store'];
+
+    $form['statement_descriptors'] = [
+      '#type' => 'details',
+      '#title' => $this->t('Statement descriptors'),
+    ];
+
+    $form['statement_descriptors']['help'] = [
+      '#markup' => $this->t(
+        'You can override the statement descriptor and the support phone number
+        appended in the descriptor here. If left empty, the default descriptor
+        and phone number configured in the BlueSnap account will be used.
+        <a href="@link" target="_blank">Learn more</a>',
+        ['@link' => 'https://support.bluesnap.com/docs/statement-descriptor']
+      ),
+    ];
+    $form['statement_descriptors']['help_2'] = [
+      '#markup' => '<p>' . $this->t(
+        'If you use tokens in statement descriptors or support phone numbers,
+        make sure that the selected values that will be replaced at the moment
+        the payment is made will always comply with the
+        <a href="@link" target="_blank">allowed characters</a>; payments for
+        which statement descriptors or support phone numbers contain illegal
+        characters or are longer than the maximum length permitted by BlueSnap
+        may fail and it may not be possible for the customers to continue and
+        place their orders.',
+        ['@link' => 'https://support.bluesnap.com/docs/statement-descriptor#statement-descriptor-format-and-content']
+      ) . '</p>',
+    ];
+    $form['statement_descriptors']['descriptor'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Soft descriptor'),
+      '#default_value' => $config['descriptor'],
+      '#element_validate' => ['token_element_validate'],
+      '#token_types' => $token_types,
+    ];
+    $form['statement_descriptors']['phone_number'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Support phone number'),
+      '#default_value' => $config['phone_number'],
```

Comment:
I think the replacing of the tokens for phone number above was an error because we're not using it here, right?

Reply:
Thanks for the catch, token validation is actually missing here I'll add it.